pep-0458.txt

```diff
@@ -527,243 +789,247 @@ How Should Metadata be Generated?
-MUST upload its release in a single transaction.  The uploaded set of files is
-called the "project transaction".  How PyPI MAY validate the files in a project
-transaction is discussed in a later section.  For now, the focus is on how PyPI
-will respond to a project transaction.
+When a new project release is uploaded to PyPI, PyPI MUST update the *bin-n*
+metadata responsible for the target files of the project release. Remember that
+target files are sorted into bins by their filename hashes. Consequentially,
+PyPI MUST update *snapshot* to account for the updated *bin-n* metadata, and
+*timestamp* to account for the updated *snapshot* metadata. These updates
+SHOULD be handled by automated processes, e.g. one or more *transaction
+processes* and one *snapshot process*.
+
+Each transaction process keeps track of a project upload, adds all new target
+files to the most recent, relevant *bin-n* metadata and informs the
+snapshot process to produce a consistent snapshot. Each project release SHOULD
+be handled in an atomic transaction, so that a given consistent snapshot
+contains all target files of a project release. However, transaction processes
+MAY be parallelized under the following constraints:
+
+- Pairs of transaction processes MUST NOT concurrently work on the same project.
+- Pairs of transaction processes MUST NOT concurrently work on projects that
+  belong to the same *bin-n* role.
+
+When a transaction process is finished updating the relevant *bin-n* metadata
+it informs the snapshot process to generate a new consistent snapshot. The
+snapshot process does so by taking the updated *bin-n* metadata, incrementing
+their respective version numbers, signing them with the *bin-n* role key(s),
+and writing them to *VERSION_NUMBER.bin-N.json*.
+
+Similarly, the snapshot process then takes the most recent *snapshot* metadata,
+updates its *bin-n* metadata version numbers, increments its own version
+number, signs it with the *snapshot* role key, and writes it to
+*VERSION_NUMBER.snapshot.json*.
+
+And finally, the snapshot process takes the most recent *timestamp* metadata,
+updates its *snapshot* metadata hash and version number, increments its own
+version number, sets a new expiration time, signs it with the *timestamp* role
+key, and writes it to *timestamp.json*.
+
+The snapshot process MUST generate consistent snapshots sequentially, reading
+the notifications received from the transaction process(es) from a
+concurrency-safe FIFO queue. Fortunately, the operation of signing is fast
+enough that this may be done a thousand or more times per second.
+
+If there are multiple files in a release, a project MAY release these files in
+separate transactions. For example, a project MAY release files for Windows in
+one transaction, and the files for Linux in another transaction. However, a project
+SHOULD release files that must belong together in order for everything to work
+in the same transaction.
```

Comment:
PyPI currently does not allow anything more than a single file per transaction, are we going to be required to change that for this PEP?

Reply:
Addressed (see https://github.com/python/peps/pull/1203#discussion_r350643470).